### PR TITLE
Add a division visitor method to SolveExpression.

### PR DESCRIPTION
**Pull request by jingpu**
Trying to merge this into the new simplifier branch because this PR requires a simplification rule of 
`(x*-4)/2 -> x*(-2)` which is not implemented in the master branch.

### CODE REVIEW

File: src/Solve.cpp [hunk 1]
```diff
@@ -348,6 +348,67 @@ class SolveExpression : public IRMutator2 {
...
+        bool b_uses_var = uses_var;
+        bool b_failed = failed;
+        uses_var = old_uses_var || a_uses_var || b_uses_var;
+        failed = old_failed || a_failed || b_failed;
+
+        const Add *add_a = a.as<Add>();
+        const Sub *sub_a = a.as<Sub>();
+        const Mul *mul_a = a.as<Mul>();
+
+        if (b_uses_var && !a_uses_var) {
```
Comment (abadams) [inline src/Solve.cpp]: div isn't commutative, so you can't swap a and b here like in the mul visitor

File: src/Solve.cpp [hunk 2]
```diff
@@ -348,6 +348,67 @@ class SolveExpression : public IRMutator2 {
...
+                // (f(x) + a) / b -> f(x) / b + a / b
+                expr = mutate(simplify(add_a->a / b) + add_a->b / b);
+            } else if (sub_a && !a_failed &&
+                       can_prove(sub_a->a / b * b == sub_a->a)) {
+                // (f(x) - a) / b -> f(x) / b - a / b
+                expr = mutate(simplify(sub_a->a / b) - sub_a->b / b);
+            } else if (mul_a && !a_failed &&
+                       can_prove(mul_a->b / b * b == mul_a->b)) {
+                // (f(x) * a) / b -> f(x) * (a / b)
+                expr = mutate(mul_a->a * (mul_a->b / b));
```
Comment (abadams) [inline src/Solve.cpp]: I don't think this one's right. Consider mul_a->a = 8x, mul_a->b = 3, b = 4

(8x * 3) / 4 is 6x, but 8x * (3/4) is zero.

Comment (jingpu) [reply src/Solve.cpp]: There is a condition `can_prove(mul_a->b / b * b == mul_a->b)` above that should rule out this case.

Comment (abadams) [reply src/Solve.cpp]: Oops, I misread that condition as being the same as the other two cases (mul_a->a)

**Comment by jingpu** [comment]
I think the build failures are pre-existing on the `metaprogrammed_simplifier_rules` branch. PTAL

**Comment by jingpu** [comment]
I moved the PR to #3229. Closing this one.

**Pull request closed without merging**